File: 24-Q01/q2.rb
#!/usr/bin/env ruby

require 'json'

f = File.open('./q1-input.txt')

l1 = []
l2 = []
f.readlines.each do |line|
  e1, e2 = line.split(' ')
  e1 = e1.to_i
  e2 = e2.to_i

  sortedIndex1 = 0
  l1.each do |x|
    if x > e1 then break end
    sortedIndex1 += 1
  end
  l1.insert(sortedIndex1, e1)

  sortedIndex2 = 0
  l2.each do |x|
    if x > e2 then break end
    sortedIndex2 += 1
  end
  l2.insert(sortedIndex2, e2)
end

p l1
p l2

sum = 0

p1 = 0
p2 = 0
while true
  val = [l1[p1], l2[p2]].min

  n1 = 0
  while l1[p1] == val
    n1 += 1
    p1 += 1
  end

  n2 = 0
  while l2[p2] == val
    n2 += 1
    p2 += 1
  end

  sum += val * n1 * n2

  if l1[p1] == nil or l2[p2] == nil
    break
  end
end

p sum

File: 24-Q19/q2.py
#!/usr/bin/env python

patterns = []
tasks = []

with open("input.txt", "r") as file:
	lines = file.readlines()

	patterns = lines[0].split(',')
	patterns = [p.strip() for p in patterns]

	tasks = [line.strip() for line in lines[2:]]

print(patterns)
print(tasks)

cache = {}
def num_arrangements(task, offset):
	if (task, offset) in cache:
		return cache[(task, offset)]

	arrangements = 0
	for pattern in patterns:
		if task[offset:] == pattern:
			arrangements += 1

		if task[offset:offset + len(pattern)] == pattern:
			arrangements += num_arrangements(task, offset + len(pattern))

	cache[(task, offset)] = arrangements
	return arrangements

sum = 0
for task in tasks:
	sum += num_arrangements(task, 0)

print("Sum: {}".format(sum))
